internal/internal_activity.go

```diff
@@ -302,98 +302,132 @@ func isActivityContext(inType reflect.Type) bool {
-	resultSize := len(values)
-
-	if resultSize < 1 || resultSize > 2 {
-		fnName, _ := getFunctionName(f)
-		return nil, fmt.Errorf(
-			"the function: %v signature returns %d results, it is expecting to return either error or (result, error)",
-			fnName, resultSize)
+func setActivityParametersIfNotExist(ctx Context) Context {
+	params := getActivityOptions(ctx)
+	var newParams ExecuteActivityOptions
+	if params != nil {
+		newParams = *params
+		if params.RetryPolicy != nil {
+			newRetryPolicy := *newParams.RetryPolicy
+			newParams.RetryPolicy = &newRetryPolicy
+		}
 	}
+	return WithValue(ctx, activityOptionsContextKey, &newParams)
+}
 
-	var result *commonpb.Payloads
+func setLocalActivityParametersIfNotExist(ctx Context) Context {
+	params := getLocalActivityOptions(ctx)
+	var newParams ExecuteLocalActivityOptions
+	if params != nil {
+		newParams = *params
+	}
+	return WithValue(ctx, localActivityOptionsContextKey, &newParams)
+}
 
-	// Parse result
-	if resultSize > 1 {
-		retValue := values[0]
+type activityEnvironmentInterceptor struct {
+	env                 *activityEnvironment
+	inboundInterceptor  ActivityInboundInterceptor
+	outboundInterceptor ActivityOutboundInterceptor
+	fn                  interface{}
+}
 
-		var ok bool
-		if result, ok = retValue.Interface().(*commonpb.Payloads); !ok {
-			if retValue.Kind() != reflect.Ptr || !retValue.IsNil() {
-				var err error
-				if result, err = encodeArg(dataConverter, retValue.Interface()); err != nil {
-					return nil, err
-				}
-			}
-		}
+func getActivityEnvironmentInterceptor(ctx context.Context) *activityEnvironmentInterceptor {
+	a := ctx.Value(activityEnvInterceptorContextKey)
+	if a == nil {
+		panic("getActivityEnvironmentInterceptor: Not an activity context")
 	}
+	return a.(*activityEnvironmentInterceptor)
+}
 
-	// Parse error.
-	errValue := values[resultSize-1]
-	if errValue.IsNil() {
-		return result, nil
+func getActivityOutboundInterceptor(ctx context.Context) ActivityOutboundInterceptor {
+	a := ctx.Value(activityInterceptorContextKey)
+	if a == nil {
+		panic("getActivityOutboundInterceptor: Not an activity context")
 	}
-	errInterface, ok := errValue.Interface().(error)
-	if !ok {
-		return nil, fmt.Errorf(
-			"failed to parse error result as it is not of error interface: %v",
-			errValue)
+	return a.(ActivityOutboundInterceptor)
+}
+
+func (a *activityEnvironmentInterceptor) Init(outbound ActivityOutboundInterceptor) error {
+	a.outboundInterceptor = outbound
+	return nil
+}
+
+func (a *activityEnvironmentInterceptor) ExecuteActivity(
+	ctx context.Context,
+	in *ExecuteActivityInput,
+) (interface{}, error) {
+	// Remove header from context
+	ctx = contextWithoutHeader(ctx)
+
+	return executeFunctionWithContext(ctx, a.fn, in.Args)
+}
+
+func (a *activityEnvironmentInterceptor) GetInfo(ctx context.Context) ActivityInfo {
+	// TODO(cretz): Accept the costs of copying everytime? See https://github.com/temporalio/proposals/pull/45/files#r732945623
```

Comment:
Need to remove this TODO, I think we accept this (it's a big copy, but meh, it was decided long before me in `activity.GetInfo` that copying this huge struct was ok and inconsistent with `workflow.GetInfo`).

Reply:
Size wise it's honestly not all that huge. Just a few kinda-long strings, which are just pointers anyway

Reply:
Agreed it's not that huge and usually I wouldn't bat an eye (I do tend to use pointers if I am not afraid of mutation and it's more than a couple words in size). However, could travel through many interceptor layers and is the type of call that may happen a lot in an activity. But there are definitely many other places for perf improvement before this...it's just something that always makes me do a double-take when writing.

Reply:
TODO removed